airbyte-scheduler/persistence/src/main/java/io/airbyte/scheduler/persistence/job_tracker/SlackNotificationClient.java

```diff
@@ -0,0 +1,115 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.scheduler.persistence.job_tracker;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableMap.Builder;
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.commons.map.MoreMaps;
+import io.airbyte.commons.resources.MoreResources;
+import io.airbyte.commons.yaml.Yamls;
+import io.airbyte.config.StandardWorkspace;
+import io.airbyte.config.persistence.ConfigNotFoundException;
+import io.airbyte.config.persistence.ConfigRepository;
+import io.airbyte.config.persistence.PersistenceConstants;
+import io.airbyte.validation.json.JsonValidationException;
+import java.io.IOException;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.util.Map;
+import java.util.Map.Entry;
+import org.apache.logging.log4j.util.Strings;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Notification client that uses Slack API for Incoming Webhook to send messages.
+ *
+ * This class also reads a resource YAML file that defines the template message to send.
+ *
+ * It is stored as a YAML so that we can easily change the structure of the JSON data expected by
+ * the API that we are posting to (and we can write multi-line strings more easily).
+ *
+ * For example, slack API expects some text message in the { "text" : "Hello World" } field...
+ *
+ * It is also a template as the content of the message can be customized with basic variables that
+ * are built in the metadata map (re-using the same as generated by the JobTracker class for now)
+ * The templating is very basic for now: it is just looking for keys from the metadata map
+ * surrounded by '<' '>' characters in the string
+ */
+public class SlackNotificationClient implements NotificationClient {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SlackNotificationClient.class);
+
+  private final HttpClient httpClient = HttpClient.newBuilder()
+      .version(HttpClient.Version.HTTP_2)
+      .build();
+  private final ConfigRepository configRepository;
+
+  public SlackNotificationClient(ConfigRepository configRepository) {
+    this.configRepository = configRepository;
+  }
+
+  @Override
+  public void notifyFailure(String action, Map<String, Object> metadata)
+      throws ConfigNotFoundException, IOException, JsonValidationException, InterruptedException {
+    metadata = MoreMaps.merge(metadata, generateMetadata(metadata));
```

Comment:
This is bad practice to reuse variables like that. That's generally why we put `final` in front of them.